File: build_files/kyth_shared/kyth_shared/system/bootc.py
"""bootc status, image reference, branch, and deployment helpers.

Pure stdlib — safe to import from CLI tools (update-watcher) without Qt.

Thin, cache-aware wrappers around ``bootc_query``/``bootc_policy``. Pure
pass-throughs of an already-public policy/query function (no added logic)
aren't re-defined here — import them from ``bootc_policy``/``bootc_query``
(also re-exported below) directly.
"""
from __future__ import annotations

from ..commands import run as run_command
from kyth_shared.system.process import BOOTC_CACHE_TTL, command_stdout, probe_cached
from kyth_shared.system import bootc_query as query
from kyth_shared.system.bootc_policy import (
    BranchCardView,
    BranchesView,
    REGISTRY,
    UpdateAvailabilityView,
    branch_display_name,
    branch_from_ref,
    branches_view,
    cancel_block_reason,
    default_phase,
    image_tag_for_channel as _raw_image_tag_for_channel,
    image_tag_for_kernel as _raw_image_tag_for_kernel,
    parse_update_phase,
    update_availability_view,
)

nested_get = query.nested_get
walk_strings = query.walk_strings
image_reference_from_status = query.image_reference_from_status
image_digest_from_status = query.image_digest_from_status


def fetch_bootc_status_text() -> str:
    return query.fetch_status_text()


def bootc_status_text() -> str:
    # Keep patchable compatibility boundaries used by probes and tests.
    return probe_cached("bootc-status-text", BOOTC_CACHE_TTL, fetch_bootc_status_text)


def fetch_bootc_status_data() -> dict | None:
    return query.fetch_status_data()


def bootc_status_data() -> dict | None:
    return probe_cached("bootc-status-data", BOOTC_CACHE_TTL, fetch_bootc_status_data)


def fetch_bootc_status_data_uncached() -> dict | None:
    return fetch_bootc_status_data()


def active_bootc_operation() -> str | None:
    return query.active_operation()


def bootc_proxy_running() -> bool:
    try:
        result = run_command(
            ["pgrep", "-f", "skopeo.*image-proxy"],
            capture_output=True, timeout=2, check=False,
        )
        return result.returncode == 0
    except Exception:
        return False


def bootc_cancel_block_reason(mode: str, phase: str) -> str:
    return cancel_block_reason(mode, phase)


def bootc_image_reference() -> str | None:
    data = bootc_status_data() or {}
    ref = query.image_reference_from_status(data)
    if ref:
        return ref
    ref = query.image_reference_from_status(data, status_output=bootc_status_text())
    if ref:
        return ref
    # Preserve the compatibility module's patchable status seams while
    # delegating the fallback query to the query adapter.
    return query.image_reference()


def current_branch() -> str | None:
    return probe_cached(
        "bootc-branch", BOOTC_CACHE_TTL,
        lambda: branch_from_ref(bootc_image_reference()),
    )


def current_kernel_flavor() -> str:
    def fetch() -> str:
        try:
            with open("/usr/share/kyth/kernel-flavor") as fh:
                flavor = fh.read().strip().lower()
                if flavor in {"fedora", "cachy"}:
                    return flavor
        except OSError:
            pass
        return "cachy" if "cachy" in command_stdout(["uname", "-r"]).lower() else "fedora"

    return probe_cached("kernel-flavor", 60.0, fetch)


def image_tag_for_channel(channel: str, flavor: str | None = None) -> str:
    return _raw_image_tag_for_channel(channel, flavor or current_kernel_flavor())


def image_tag_for_kernel(flavor: str) -> str:
    return _raw_image_tag_for_kernel(flavor, current_branch())


def has_staged_update() -> bool:
    return nested_get(bootc_status_data() or {}, ("status", "staged")) is not None


def has_rollback_deployment() -> bool:
    return nested_get(bootc_status_data() or {}, ("status", "rollback")) is not None


def bootc_image_timestamp(section: str) -> str | None:
    # Query through the compatibility status function so existing cache/test
    # injection remains effective.
    data = bootc_status_data() or {}
    deployment = nested_get(data, ("status", section)) or {}
    from datetime import datetime

    for path in (("image", "timestamp"), ("timestamp",)):
        value = nested_get(deployment, path)
        if isinstance(value, str) and value.strip():
            try:
                parsed = datetime.fromisoformat(value.strip().replace("Z", "+00:00")).astimezone()
                return parsed.strftime("%Y-%m-%d %H:%M %Z")
            except (ValueError, OverflowError):
                return value.strip()
    return None


def bootc_image_digest(section: str) -> tuple[str, str] | None:
    value = image_digest_from_status(bootc_status_data(), section)
    return None if value is None else (value[7:19], value[7:])

File: src/kyth-shared-rs/src/system/bootc.rs
//! Port of `kyth_shared.system.bootc` — thin cache wrappers around bootc_query/policy.

pub fn branch_from_ref(r: Option<&str>) -> Option<String> {
    crate::system::bootc_policy::branch_from_ref(r)
}

pub fn current_branch() -> Option<String> {
    // Prefer the probe cache, then mirror Python's cache-miss behavior by
    // deriving the branch from the current bootc status response.
    if let Some(branch) = crate::system::probe::read_section("bootc-branch")
        .and_then(|v| v.as_str().map(str::to_string))
    {
        return Some(branch);
    }
    let status = crate::system::probe::read_section("bootc-status-data")
        .or_else(crate::system::bootc_query::fetch_status_data)?;
    crate::system::bootc_query::image_reference_from_status(&status)
        .as_deref()
        .and_then(|reference| branch_from_ref(Some(reference)))
}

pub fn current_kernel_flavor() -> String {
    if let Ok(s) = std::fs::read_to_string("/usr/share/kyth/kernel-flavor") {
        let f = s.trim().to_lowercase();
        if f == "cachy" || f == "fedora" { return f; }
    }
    // fallback uname -r check
    if let Some((_, stdout)) = run_with_timeout(&["uname".to_string(), "-r".to_string()], std::time::Duration::from_secs(2)) {
        if stdout.to_lowercase().contains("cachy") { return "cachy".to_string(); }
    }
    "fedora".to_string()
}

fn run_with_timeout(cmd: &[String], timeout: std::time::Duration) -> Option<(i32, String)> {
    if cmd.is_empty() { return None; }
    let output = super::process::run_bounded(cmd, timeout).ok()?;
    Some((output.status.code().unwrap_or(-1), String::from_utf8_lossy(&output.stdout).to_string()))
}

pub fn has_staged_update() -> bool {
    crate::system::probe::read_section("bootc-status-data")
        .or_else(crate::system::bootc_query::fetch_status_data)
        .and_then(|v| v.get("status").and_then(|s| s.get("staged")).cloned())
        .is_some()
}

pub fn has_rollback_deployment() -> bool {
    crate::system::probe::read_section("bootc-status-data")
        .or_else(crate::system::bootc_query::fetch_status_data)
        .and_then(|v| v.get("status").and_then(|s| s.get("rollback")).cloned())
        .is_some()
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn branch() {
        assert_eq!(branch_from_ref(Some("ghcr.io/kyth-os/kyth:latest")), Some("latest".to_string()));
    }
    #[test]
    fn staged_bool() {
        let _ = has_staged_update();
    }
}
